src/backend/distributed/commands/extension.c

```diff
@@ -113,5 +109,523 @@ ExtractNewExtensionVersion(Node *parsetree)
+ */
+static void
+AddSchemaFieldIfMissing(CreateExtensionStmt *createExtensionStmt)
+{
+	List *optionsList = createExtensionStmt->options;
+
+	const char *schemaName = GetCreateExtensionOption(optionsList, "schema");
+
+	if (!schemaName)
+	{
+		char *currentSchemaName = GetCurrentSchema();
+
+		Node *schemaNameArg = (Node *) makeString(currentSchemaName);
+
+		/* set location to -1 as it is unknown */
+		int location = -1;
+
+		DefElem *newDefElement = makeDefElem("schema", schemaNameArg, location);
+
+		createExtensionStmt->options = lappend(createExtensionStmt->options,
+											   newDefElement);
+	}
+}
+
+
+/*
+ * Utility function to fetch the string value of DefElem node with "defname"
+ * from "options" list
+ */
+const char *
+GetCreateExtensionOption(List *extensionOptions, const char *defname)
+{
+	const char *targetStr = NULL;
+
+	ListCell *defElemCell = NULL;
+
+	foreach(defElemCell, extensionOptions)
+	{
+		DefElem *defElement = (DefElem *) lfirst(defElemCell);
+
+		if (IsA(defElement, DefElem) && strncmp(defElement->defname, defname,
+												NAMEDATALEN) == 0)
+		{
+			targetStr = strVal(defElement->arg);
+			break;
+		}
+	}
+
+	/* return target string safely */
+	if (targetStr)
+	{
+		return pstrdup(targetStr);
+	}
+	else
+	{
+		return NULL;
+	}
+}
+
+
+/*
+ * Get the name of the schema that the postgres would pick primarily
+ * for a CREATE EXTENSION statement that does not include "WITH SCHEMA"
+ * clause.
+ * Cannot return NULL, errors out as Postgres would do if NULL.
+ */
+static char *
+GetCurrentSchema(void)
+{
+	/*
+	 * Neither user nor author of the extension specified schema; use the
+	 * current default creation namespace, which is the first explicit
+	 * entry in the search_path.
+	 */
+	List *search_path = fetch_search_path(false);
+	Oid schemaOid = InvalidOid;
+	char *schemaName = NULL;
+
+	/* nothing valid in search_path? */
+	if (search_path == NIL)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("no schema has been selected to create in")));
+	}
+	schemaOid = linitial_oid(search_path);
+	schemaName = get_namespace_name(schemaOid);
+
+	/* recently-deleted namespace? */
+	if (schemaName == NULL)
+	{
+		ereport(ERROR,
+				(errcode(ERRCODE_UNDEFINED_SCHEMA),
+				 errmsg("no schema has been selected to create in")));
+	}
+
+	list_free(search_path);
+
+	return pstrdup(schemaName);
+}
+
+
+/*
+ * ProcessCreateExtensionStmt is executed after the extension has been
+ * created locally and before we create it on the worker nodes.
+ * As we now have access to ObjectAddress of the extension that is just
+ * created, we can mark it as distributed to make sure that its
+ * dependencies exist on all nodes.
+ */
+void
+ProcessCreateExtensionStmt(CreateExtensionStmt *createExtensionStmt, const
+						   char *queryString)
+{
+	const ObjectAddress *extensionAddress = NULL;
+
+	if (!ShouldPropagateExtensionCommand((Node *) createExtensionStmt))
+	{
+		return;
+	}
+
```

Comment:
To prevent concurrent node addition, we should have something like this here:
```C
        /* don't allow concurrent node list changes that require an exclusive lock */
        LockRelationOid(DistNodeRelationId(), RowShareLock);
```

Otherwise, in one session, you can add a node, and on the other you can add an extension, and that the extension might not get propagated to the newly added node.


We should verify this  with an  isolation test

Reply:
Implemented in bddad5780f18cfc01cba610e3f0c78175372cb93, I will add related isolation tests soon